// @bone/sdk-base/lib/interface/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hub_1 = require("../hub");
// module name for hub
exports.MODULE_NAME_CHANNEL = 'channel';
// event names
exports.EVENT_CONNECT_STATUS_CHANGE = 'BoneChannelConnectStatusChange';
exports.EVENT_MESSAGE = 'BoneChannelMessage';
var CHANNEL_CONNECT_STATUS;
(function (CHANNEL_CONNECT_STATUS) {
    CHANNEL_CONNECT_STATUS["CONNECTED"] = "connected";
    CHANNEL_CONNECT_STATUS["DISCONNECTED"] = "disconnected";
})(CHANNEL_CONNECT_STATUS = exports.CHANNEL_CONNECT_STATUS || (exports.CHANNEL_CONNECT_STATUS = {}));
function ready(config) {
    return hub_1.call(exports.MODULE_NAME_CHANNEL, 'ready', [config]);
}
exports.ready = ready;
;
function subTopic(topic, onMessage) {
    hub_1.call(exports.MODULE_NAME_CHANNEL, 'subTopic', [topic, onMessage]);
}
exports.subTopic = subTopic;
;
function unsubTopic(topic, onMessage) {
    hub_1.call(exports.MODULE_NAME_CHANNEL, 'unsubTopic', [topic, onMessage]);
}
exports.unsubTopic = unsubTopic;
;
function publish(topic, payload) {
    return hub_1.call(exports.MODULE_NAME_CHANNEL, 'publish', [topic, payload]);
}
exports.publish = publish;
;
function request(topic, payload) {
    return hub_1.call(exports.MODULE_NAME_CHANNEL, 'request', [topic, payload]);
}
exports.request = request;
;
function onMessage(callback) {
    hub_1.message.on(exports.EVENT_MESSAGE, callback);
}
exports.onMessage = onMessage;
;
function onConnectStatusChange(callback) {
    hub_1.message.on(exports.EVENT_CONNECT_STATUS_CHANGE, callback);
}
exports.onConnectStatusChange = onConnectStatusChange;
;
