// @bone/sdk-base/lib/interface/netInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hub_1 = require("../hub");
var constants_1 = require("./constants");
// @todo
// 各个页面监听，需要隔离
var NETWORK_STATUS;
(function (NETWORK_STATUS) {
    // 有网
    NETWORK_STATUS["WIFI"] = "wifi";
    NETWORK_STATUS["CELL"] = "cell";
    NETWORK_STATUS["OTHERS"] = "others";
    // 无网
    NETWORK_STATUS["NONE"] = "none";
    // 未知
    NETWORK_STATUS["UNKNOWN"] = "unknown";
})(NETWORK_STATUS = exports.NETWORK_STATUS || (exports.NETWORK_STATUS = {}));
exports.EVENT_NETINFO_CHANGE = 'netInfoChanged';
function fetch() {
    return hub_1.call(constants_1.MODULE_NAME_HUB.NETWORK, 'fetchNetInfo');
}
exports.fetch = fetch;
;
function isConnected() {
    return new Promise(function (resolve, reject) {
        fetch()
            .then(function (status) { return resolve(status !== NETWORK_STATUS.NONE); });
    });
}
exports.isConnected = isConnected;
function onChange(callback) {
    hub_1.message.on(exports.EVENT_NETINFO_CHANGE, callback);
}
exports.onChange = onChange;
;
