// @bone/sdk-base/lib/interface/sysInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hub_1 = require("../hub");
var constants_1 = require("./constants");
// runtime	运行环境	String	N	“bone-rn”		RN容器中，固定为"BoneReactNative"
// apiLevel	运行环境版本	String	N	“2.0.0”		
// pluginEnv	面板环境	String	N	“release”	“release”,“test”,“alpha”	
// serverEnv	网络环境	String	N	“online”	“online”,“preview”,“daily”	
// appName	app名称	String	N	“阿里小智”		
// appVersion	app版本	String	N	“2.0.0”		
// phone	手机型号	String	N	“XiaoMi M6”		
// platform	手机系统平台	String	N	“Android”	“Android”,“iOS”	
// systemVersion	系统版本	String	N	“10.2”		备注：Android的版本号请参考下文
var RUNTIME;
(function (RUNTIME) {
    RUNTIME["RN"] = "bone-rn";
    // @todo
    // 还有哪些runtime
    // runtime定义是否要更新
})(RUNTIME = exports.RUNTIME || (exports.RUNTIME = {}));
var ENV;
(function (ENV) {
    ENV["PRODUCTION"] = "production";
    ENV["TEST"] = "test";
    ENV["DEVELOPMENT"] = "development";
})(ENV = exports.ENV || (exports.ENV = {}));
var PLATFORM;
(function (PLATFORM) {
    PLATFORM["iOS"] = "ios";
    PLATFORM["Android"] = "android";
    PLATFORM["Window"] = "window";
    PLATFORM["Linux"] = "linux";
    PLATFORM["Others"] = "others";
})(PLATFORM = exports.PLATFORM || (exports.PLATFORM = {}));
function fetch() {
    return hub_1.call(constants_1.MODULE_NAME_HUB.SYS, 'fetch');
}
exports.fetch = fetch;
