// @bone/bone-mobile-ui/lib/ItemGrid/lib/constant.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var NOOP = function NOOP() {};

var RECT = 'rect',
    CIRCLE = 'circle';


var RECT_VIEW_HEIGHT = 85;

var CIRCLE_VIEW_WIDTH = 68,
    CIRCLE_VIEW_HEIGHT = 90;
exports.NOOP = NOOP;
exports.RECT = RECT;
exports.CIRCLE = CIRCLE;
exports.RECT_VIEW_HEIGHT = RECT_VIEW_HEIGHT;
exports.CIRCLE_VIEW_WIDTH = CIRCLE_VIEW_WIDTH;
exports.CIRCLE_VIEW_HEIGHT = CIRCLE_VIEW_HEIGHT;
var _ref = [Symbol('react-native-item-grid-key'), Symbol('react-native-item-grid-status'), Symbol('react-native-item-grid-index'), Symbol('react-native-item-grid-text'), Symbol('react-native-item-grid-icon'), Symbol('react-native-item-grid-value')],
    keySymbol = _ref[0],
    statusSymbol = _ref[1],
    indexSymbol = _ref[2],
    textSymbol = _ref[3],
    iconSymbol = _ref[4],
    valueSymbol = _ref[5];
exports.keySymbol = keySymbol;
exports.statusSymbol = statusSymbol;
exports.indexSymbol = indexSymbol;
exports.textSymbol = textSymbol;
exports.iconSymbol = iconSymbol;
exports.valueSymbol = valueSymbol;
