// @bone/bone-mobile-ui/lib/RootModal/lib/containers/ToastContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _Container2 = require('./Container');

var _Container3 = _interopRequireDefault(_Container2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var windowInnerHeight = _reactNative.Dimensions.get('window').height;

var duration = 100;

var ToastContainer = function (_Container) {
    _inherits(ToastContainer, _Container);

    function ToastContainer() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, ToastContainer);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ToastContainer.__proto__ || Object.getPrototypeOf(ToastContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
            visible: false,
            opacity: new _reactNative.Animated.Value(0)
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(ToastContainer, [{
        key: '$stopAnimate',
        value: function $stopAnimate() {
            var opacity = this.state.opacity;


            opacity.stopAnimation();
        }

        // 重定义 show 函数

    }, {
        key: 'show',
        value: function show() {
            var opacity = this.state.opacity;


            this.$stopAnimate();

            this.setState({
                visible: true
            });

            _reactNative.Animated.timing(opacity, {
                duration: duration,
                toValue: 1,
                easing: _reactNative.Easing.linear
            }).start();
        }

        // 重定义 hide 函数

    }, {
        key: 'hide',
        value: function hide() {
            var opacity = this.state.opacity;


            this.$stopAnimate();

            this.setState({
                visible: false
            });

            opacity.setValue(0);

            this.props.onHideComplete();
        }

        // 重定义 render 函数

    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                onBackgroundPress = _props.onBackgroundPress,
                children = _props.children;
            var _state = this.state,
                visible = _state.visible,
                opacity = _state.opacity;


            return visible ? _react2.default.createElement(
                _reactNative.View,
                { style: styles.body },
                _react2.default.createElement(
                    _reactNative.View,
                    { style: styles.background },
                    _react2.default.createElement(_reactNative.TouchableOpacity, {
                        style: styles.backgroundTouch,
                        activeOpacity: 1,
                        onPress: onBackgroundPress
                    })
                ),
                _react2.default.createElement(
                    _reactNative.Animated.View,
                    { style: { opacity: opacity } },
                    children
                )
            ) : null;
        }
    }]);

    return ToastContainer;
}(_Container3.default);

exports.default = ToastContainer;


var styles = _reactNative.StyleSheet.create({
    body: {
        position: 'absolute',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        paddingTop: windowInnerHeight / 667 * 222,
        backgroundColor: 'transparent',
        overflow: 'hidden',
        alignItems: 'center'
    },
    background: {
        position: 'absolute',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        backgroundColor: 'transparent'
    },
    backgroundTouch: {
        flex: 1
    }
});
