// @bone/bone-mobile-ui/lib/ItemGrid/lib/Item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _constant = require('./constant');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// 保留 ^1.0.0 版本的接口
var Item = function () {
    function Item(_ref) {
        var key = _ref.key,
            status = _ref.status,
            index = _ref.index,
            text = _ref.text,
            icon = _ref.icon,
            value = _ref.value;

        _classCallCheck(this, Item);

        this[_constant.keySymbol] = key;

        this[_constant.statusSymbol] = status;

        this[_constant.indexSymbol] = index;

        this[_constant.textSymbol] = text;

        this[_constant.iconSymbol] = icon;

        this[_constant.valueSymbol] = value;
    }

    // getters


    _createClass(Item, [{
        key: 'isDisable',


        // boolean funtions
        // 忽略了 disabled 属性，和 ^1.0.0 版本保持一致
        value: function isDisable() {

            return this.status === -1;
        }
    }, {
        key: 'isNormal',
        value: function isNormal() {

            return this.status === 0;
        }
    }, {
        key: 'isHighlight',
        value: function isHighlight() {

            return this.status === 1;
        }

        // action functions
        // 提供通道，直接修改组件内部状态，而不去同步外部状态
        // 接口不太合理，和 ^1.0.0 版本保持一致，但不实现

    }, {
        key: 'disabled',
        value: function disabled() {}
    }, {
        key: 'normal',
        value: function normal() {}
    }, {
        key: 'highlight',
        value: function highlight() {}
    }, {
        key: 'key',
        get: function get() {

            return this[_constant.keySymbol];
        }
    }, {
        key: 'status',
        get: function get() {

            return this[_constant.statusSymbol];
        }
    }, {
        key: 'index',
        get: function get() {

            return this[_constant.indexSymbol];
        }
    }, {
        key: 'text',
        get: function get() {

            return this[_constant.textSymbol];
        }
    }, {
        key: 'icon',
        get: function get() {

            return this[_constant.iconSymbol];
        }
    }, {
        key: 'value',
        get: function get() {

            return this[_constant.valueSymbol];
        }
    }]);

    return Item;
}();

exports.default = Item;
