// @bone/bone-mobile/lib/globals.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isProduction = exports.env = exports.size = exports.OS = exports.isAlipay = exports.isWeb = exports.isRN = undefined;

var _reactNative = require('react-native');

var env = process ? process.env.NODE_ENV : 'production';

var isProduction = env === 'production';

var OS = _reactNative.Platform.OS; // iOS/android

var isWeb = OS === 'web';

var isRN = !isWeb;

var isAlipay = isWeb && !!window.AlipayJSBridge;

// @todo isTaobao
var size = {
  // react native ios: 64
  // react native android: 44
  // web simulator: 44
  // @todo nav height calculation in alipay
  navHeight: _reactNative.Platform.OS === 'ios' ? 64 : 44,

  get height() {
    // 魅族2 兼容 smartbar
    var height = _reactNative.Dimensions.get('window').height;

    // @todo
    // var phone = Bone.info.phone || "";
    // if(phone && phone.indexOf('mx2') >= 0){
    //     height -= 50;
    // }
    return height;
  },

  get width() {
    return _reactNative.Dimensions.get('window').width;
  }
};

exports.isRN = isRN;
exports.isWeb = isWeb;
exports.isAlipay = isAlipay;
exports.OS = OS;
exports.size = size;
exports.env = env;
exports.isProduction = isProduction;
