// @bone/sdk-base/lib/implement/RN/sysInfo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("./constants");
var sysInfo_1 = require("../../interface/sysInfo");
var bridge_1 = __importDefault(require("./bridge"));
var pluginEnvMapping = {
    release: sysInfo_1.ENV.PRODUCTION,
    test: sysInfo_1.ENV.TEST,
    alpha: sysInfo_1.ENV.DEVELOPMENT,
};
var serverEnvMapping = {
    production: sysInfo_1.ENV.PRODUCTION,
    test: sysInfo_1.ENV.TEST,
    development: sysInfo_1.ENV.DEVELOPMENT,
};
function fetch() {
    return new Promise(function (resolve, reject) {
        bridge_1.default
            .call(constants_1.MODULE_NAME_NATIVE.SYS, 'getSystemInfo', [], function (err, rawInfo) {
            if (err) {
                reject(err);
            }
            else {
                var phone = rawInfo.phone, systemVersion = rawInfo.systemVersion, language = rawInfo.language, platform = rawInfo.platform, 
                // @deprecated
                // @see containerInfo
                runtime = rawInfo.runtime, pluginEnv = rawInfo.pluginEnv, serverEnv = rawInfo.serverEnv;
                resolve({
                    userAgent: phone,
                    systemVersion: systemVersion,
                    language: language,
                    platform: platform,
                    // @deprecated
                    // @see containerInfo
                    runtime: runtime,
                    pluginEnv: pluginEnvMapping[pluginEnv] || sysInfo_1.ENV.PRODUCTION,
                    serverEnv: serverEnvMapping[serverEnv] || sysInfo_1.ENV.PRODUCTION,
                });
            }
        });
    });
}
exports.fetch = fetch;
