// @bone/bone-mobile-ui/lib/RootSiblings/lib/RootSiblings.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

var _AppRegistryInjection = require('./AppRegistryInjection');

var _AppRegistryInjection2 = _interopRequireDefault(_AppRegistryInjection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var uid = 0;

var RootSiblings = function () {
    function RootSiblings(element, callback) {
        _classCallCheck(this, RootSiblings);

        this.$id = null;


        Object.defineProperty(this, '$id', {
            enumerable: false,
            configurable: false,
            writable: false,
            value: uid++
        });

        this.update(element, callback);
    }

    _createClass(RootSiblings, [{
        key: '$offStreamElement',
        value: function $offStreamElement(element) {

            return (0, _react.cloneElement)(element, {
                style: [element.props.style, styles.offStream]
            });
        }
    }, {
        key: 'update',
        value: function update(element, callback) {

            // 兼容 Bone V1
            if (typeof Bone !== 'undefined' && Bone.AppRegistryInjection && Bone.AppRegistryInjection.emit) {

                Bone.AppRegistryInjection.emit('bone.ar.update', this.$id, this.$offStreamElement(element), callback);
            } else {

                _AppRegistryInjection2.default.update(this.$id, this.$offStreamElement(element), callback);
            }
        }
    }, {
        key: 'destroy',
        value: function destroy(callback) {

            // 兼容 Bone V1
            if (typeof Bone !== 'undefined' && Bone.AppRegistryInjection && Bone.AppRegistryInjection.emit) {

                Bone.AppRegistryInjection.emit('bone.ar.update', this.$id, null, callback);
            } else {

                _AppRegistryInjection2.default.update(this.$id, null, callback);
            }
        }
    }]);

    return RootSiblings;
}();

exports.default = RootSiblings;


var styles = _reactNative.StyleSheet.create({
    offStream: {
        position: 'absolute'
    }
});
