// @bone/sdk-base/lib/interface/device/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// basic APIs
var APIs_1 = require("./APIs");
exports.DEVICE_STATUS = APIs_1.DEVICE_STATUS;
exports.BoneDeviceEvent = APIs_1.BoneDeviceEvent;
exports.BoneDeviceStatusChange = APIs_1.BoneDeviceStatusChange;
exports.BoneDevicePropertiesChange = APIs_1.BoneDevicePropertiesChange;
exports.subscribeAll = APIs_1.subscribeAll;
exports.unsubscribeAll = APIs_1.unsubscribeAll;
exports.getStatus = APIs_1.getStatus;
exports.getProps = APIs_1.getProps;
exports.setProps = APIs_1.setProps;
exports.getEvents = APIs_1.getEvents;
exports.callService = APIs_1.callService;
exports.onEvent = APIs_1.onEvent;
exports.onStatusChange = APIs_1.onStatusChange;
exports.onPropsChange = APIs_1.onPropsChange;
// advanced class-style implement
var Device_1 = __importDefault(require("./Device"));
exports.Device = Device_1.default;
