// @bone/bone-mobile-ui/lib/PlaceHolder/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _TextRow = require('./TextRow');

var _TextRow2 = _interopRequireDefault(_TextRow);

var _RoundShape = require('./RoundShape');

var _RoundShape2 = _interopRequireDefault(_RoundShape);

var _RectShape = require('./RectShape');

var _RectShape2 = _interopRequireDefault(_RectShape);

var _TextBlock = require('./TextBlock');

var _TextBlock2 = _interopRequireDefault(_TextBlock);

var _MediaBlock = require('./MediaBlock');

var _MediaBlock2 = _interopRequireDefault(_MediaBlock);

var _TabShap = require('./TabShap');

var _TabShap2 = _interopRequireDefault(_TabShap);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import Card from 'Card';

exports.default = {
    textRow: _TextRow2.default,
    round: _RoundShape2.default,
    rect: _RectShape2.default,
    text: _TextBlock2.default,
    media: _MediaBlock2.default,
    tab: _TabShap2.default,
    TextRow: _TextRow2.default,
    RoundShape: _RoundShape2.default,
    RectShape: _RectShape2.default,
    TextBlock: _TextBlock2.default,
    MediaBlock: _MediaBlock2.default,
    TabShap: _TabShap2.default
    // card:  Card
};
