// @bone/bone-mobile-ui/lib/PlaceHolder/lib/TabShap.js
/**
 * @akModule AKToast
 * @desciption
 * @aliplus.com
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactNative = require('react-native');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TabShap = function (_React$Component) {
  _inherits(TabShap, _React$Component);

  function TabShap() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TabShap);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TabShap.__proto__ || Object.getPrototypeOf(TabShap)).call.apply(_ref, [this].concat(args))), _this), _this.getRowStyle = function (i) {
      var _this$props = _this.props,
          rows = _this$props.rows,
          color = _this$props.color;


      var style = {
        backgroundColor: color,
        flex: 1
      };
      return style;
    }, _this.getRows = function () {
      var rows = _this.props.rows * 2 - 1;
      var range = Array.apply(null, { length: rows });
      return range.map(function (x, i) {
        return _react2.default.createElement(TextRow, { style: _this.getRowStyle(i), invisible: i % 2 == 0, key: i });
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TabShap, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          style = _props.style,
          className = _props.className;


      return _react2.default.createElement(
        _reactNative.View,
        { style: _extends({}, style, { flex: 1, flexDirection: 'column' }) },
        this.getRows()
      );
    }
  }]);

  return TabShap;
}(_react2.default.Component);

TabShap.propTypes = {
  rows: _react2.default.PropTypes.number.isRequired,
  color: _react2.default.PropTypes.string.isRequired
};
exports.default = TabShap;
